File: cli/lib/proxy.js
const { getDb, saveDb } = require('./store');

function toggleProxy(enabled, jsonMode = false) {
  const db = getDb();
  if (enabled === undefined) {
    db.settings.isProxyModeEnabled = !db.settings.isProxyModeEnabled;
  } else {
    db.settings.isProxyModeEnabled = Boolean(enabled);
  }
  saveDb(db);

  const statusStr = db.settings.isProxyModeEnabled ? `⚡ PROXY LOW-RES ENABLED (${db.settings.proxyResolution})` : "🎬 ORIGINAL 1080p FULL-RES ENABLED";
  if (jsonMode) return JSON.stringify({ success: true, isProxyModeEnabled: db.settings.isProxyModeEnabled, proxyResolution: db.settings.proxyResolution }, null, 2);
  return `\n🔄 Mode Pratinjau Timeline Diubah: ${statusStr}\n`;
}

function setProxyResolution(resolution, jsonMode = false) {
  const db = getDb();
  db.settings.proxyResolution = resolution;
  if (resolution.includes("1080p") || resolution.includes("Original")) {
    db.settings.isProxyModeEnabled = false;
  } else {
    db.settings.isProxyModeEnabled = true;
  }
  saveDb(db);

  if (jsonMode) return JSON.stringify({ success: true, proxyResolution: db.settings.proxyResolution, isProxyModeEnabled: db.settings.isProxyModeEnabled }, null, 2);
  return `\n⚡ Target Resolusi Proxy Diperbarui Ke: "${resolution}". Mode Proxy: ${db.settings.isProxyModeEnabled ? "AKTIF" : "NONAKTIF"}\n`;
}

function toggleAutoTranscode(enabled, jsonMode = false) {
  const db = getDb();
  if (enabled === undefined) {
    db.settings.autoTranscodeOnImport = !db.settings.autoTranscodeOnImport;
  } else {
    db.settings.autoTranscodeOnImport = Boolean(enabled);
  }
  saveDb(db);

  if (jsonMode) return JSON.stringify({ success: true, autoTranscodeOnImport: db.settings.autoTranscodeOnImport }, null, 2);
  return `\n🔄 Auto-Transcode Impor Aset: ${db.settings.autoTranscodeOnImport ? "AKTIF (Otomatis konversi video baru di latar belakang)" : "NONAKTIF"}\n`;
}

function transcodeClip(clipId, jsonMode = false) {
  const db = getDb();
  const cId = parseInt(clipId);
  const clip = db.clips.find(c => c.id === cId);
  if (!clip) {
    if (jsonMode) return JSON.stringify({ error: "Clip not found" });
    return `❌ Error: Klip ID ${clipId} tidak ditemukan.`;
  }

  const targetRes = db.settings.proxyResolution;
  clip.proxyStatus = "READY";
  clip.proxyUri = `proxy_${targetRes.take ? targetRes.take(4).toLowerCase() : '360p'}_${clip.id}.mp4`;

  const job = {
    id: "job_manual_" + Date.now(),
    clipId: clip.id,
    mediaTitle: clip.title,
    originalResolution: "1080p FHD",
    targetResolution: targetRes,
    progressPercent: 100,
    statusMessage: `Proxy Low-Res Transcoded (${targetRes})`,
    isCompleted: true
  };

  db.transcodingJobs.push(job);
  saveDb(db);

  if (jsonMode) return JSON.stringify({ success: true, clip, job }, null, 2);
  return `\n⚡ Klip ID ${clipId} ("${clip.title}") berhasil ditranscode ke ${targetRes}! Status: READY.\n`;
}

function transcodeAll(jsonMode = false) {
  const db = getDb();
  const activeProjectId = db.activeProjectId;
  const projectClips = db.clips.filter(c => c.projectId === activeProjectId);

  const targetRes = db.settings.proxyResolution;
  let count = 0;

  projectClips.forEach(clip => {
    clip.proxyStatus = "READY";
    clip.proxyUri = `proxy_${clip.id}.mp4`;
    count++;

    db.transcodingJobs.push({
      id: "job_all_" + clip.id + "_" + Date.now(),
      clipId: clip.id,
      mediaTitle: clip.title,
      originalResolution: "1080p FHD",
      targetResolution: targetRes,
      progressPercent: 100,
      statusMessage: `Proxy Transcoded (${targetRes})`,
      isCompleted: true
    });
  });

  saveDb(db);

  if (jsonMode) return JSON.stringify({ success: true, transcodedCount: count, targetResolution: targetRes }, null, 2);
  return `\n🚀 BACKGROUND TRANSCODER ENGINE\n=========================================\n✅ Berhasil mentranscode ${count} klip video ke resolusi proxy (${targetRes})!\nBeban GPU Timeline berkurang ~75% untuk pemutaran 60 FPS tanpa lag.\n`;
}

function viewProxyStatus(jsonMode = false) {
  const db = getDb();
  const s = db.settings;
  const jobs = db.transcodingJobs;

  if (jsonMode) return JSON.stringify({ settings: s, jobs }, null, 2);

  let out = `\n⚙️ STATUS LOW-RESOLUTION PROXY ENGINE\n`;
  out += `=========================================\n`;
  out += `Mode Proxy Preview : ${s.isProxyModeEnabled ? "⚡ AKTIF" : "🎬 NONAKTIF (1080p Direct)"}\n`;
  out += `Target Resolution  : ${s.proxyResolution}\n`;
  out += `Auto-Transcode     : ${s.autoTranscodeOnImport ? "AKTIF" : "NONAKTIF"}\n`;
  out += `Beban GPU Estimasi : ${s.isProxyModeEnabled ? "15% - 25% (Sangat Ringan)" : "85% - 100% (High-Res Direct)"}\n\n`;

  out += `📋 RIWAYAT TUGAS TRANSCODER LATAR BELAKANG (${jobs.length} Tugas):\n`;
  if (jobs.length === 0) {
    out += `   (Belum ada tugas transcoding)\n`;
  } else {
    jobs.slice(-5).forEach(j => {
      out += `   - [${j.isCompleted ? "DONE" : "IN_PROGRESS"}] ${j.mediaTitle} (${j.originalResolution} -> ${j.targetResolution}) - Progress: ${j.progressPercent}%\n`;
    });
  }
  return out;
}

module.exports = {
  toggleProxy,
  setProxyResolution,
  toggleAutoTranscode,
  transcodeClip,
  transcodeAll,
  viewProxyStatus
};

File: cli/lib/proxy.py
import json
import time
from .store import get_db, save_db

def toggle_proxy(enabled=None, json_mode=False):
    db = get_db()
    settings = db.get("settings", {})
    if enabled is None:
        settings["isProxyModeEnabled"] = not settings.get("isProxyModeEnabled", True)
    else:
        settings["isProxyModeEnabled"] = bool(enabled)
    db["settings"] = settings
    save_db(db)

    res = settings.get("proxyResolution", "360p Proxy")
    status_str = f"⚡ PROXY LOW-RES ENABLED ({res})" if settings["isProxyModeEnabled"] else "🎬 ORIGINAL 1080p FULL-RES ENABLED"
    if json_mode:
        return json.dumps({"success": True, "isProxyModeEnabled": settings["isProxyModeEnabled"], "proxyResolution": res}, indent=2)
    return f"\n🔄 Mode Pratinjau Timeline Diubah: {status_str}\n"

def set_proxy_resolution(resolution, json_mode=False):
    db = get_db()
    settings = db.get("settings", {})
    settings["proxyResolution"] = resolution
    if "1080p" in resolution or "Original" in resolution:
        settings["isProxyModeEnabled"] = False
    else:
        settings["isProxyModeEnabled"] = True

    db["settings"] = settings
    save_db(db)

    if json_mode:
        return json.dumps({"success": True, "proxyResolution": settings["proxyResolution"], "isProxyModeEnabled": settings["isProxyModeEnabled"]}, indent=2)
    return f"\n⚡ Target Resolusi Proxy Diperbarui Ke: \"{resolution}\". Mode Proxy: {'AKTIF' if settings['isProxyModeEnabled'] else 'NONAKTIF'}\n"

def toggle_auto_transcode(enabled=None, json_mode=False):
    db = get_db()
    settings = db.get("settings", {})
    if enabled is None:
        settings["autoTranscodeOnImport"] = not settings.get("autoTranscodeOnImport", True)
    else:
        settings["autoTranscodeOnImport"] = bool(enabled)

    db["settings"] = settings
    save_db(db)

    if json_mode:
        return json.dumps({"success": True, "autoTranscodeOnImport": settings["autoTranscodeOnImport"]}, indent=2)
    return f"\n🔄 Auto-Transcode Impor Aset: {'AKTIF (Otomatis konversi video baru di latar belakang)' if settings['autoTranscodeOnImport'] else 'NONAKTIF'}\n"

def transcode_clip(clip_id, json_mode=False):
    db = get_db()
    try:
        cid = int(clip_id)
    except (ValueError, TypeError):
        cid = -1

    clip = next((c for c in db.get("clips", []) if c.get("id") == cid), None)
    if not clip:
        if json_mode:
            return json.dumps({"error": "Clip not found"})
        return f"❌ Error: Klip ID {clip_id} tidak ditemukan."

    target_res = db.get("settings", {}).get("proxyResolution", "360p Proxy")
    res_tag = target_res[:4].lower() if target_res else "360p"
    clip["proxyStatus"] = "READY"
    clip["proxyUri"] = f"proxy_{res_tag}_{clip['id']}.mp4"

    job = {
        "id": f"job_manual_{int(time.time() * 1000)}",
        "clipId": clip["id"],
        "mediaTitle": clip["title"],
        "originalResolution": "1080p FHD",
        "targetResolution": target_res,
        "progressPercent": 100,
        "statusMessage": f"Proxy Low-Res Transcoded ({target_res})",
        "isCompleted": True
    }

    db.setdefault("transcodingJobs", []).append(job)
    save_db(db)

    if json_mode:
        return json.dumps({"success": True, "clip": clip, "job": job}, indent=2)
    clip_title = clip['title']
    return f"\n⚡ Klip ID {clip_id} (\"{clip_title}\") berhasil ditranscode ke {target_res}! Status: READY.\n"

def transcode_all(json_mode=False):
    db = get_db()
    active_project_id = db.get("activeProjectId")
    project_clips = [c for c in db.get("clips", []) if c.get("projectId") == active_project_id]
    target_res = db.get("settings", {}).get("proxyResolution", "360p Proxy")
    count = 0

    for clip in project_clips:
        clip["proxyStatus"] = "READY"
        clip["proxyUri"] = f"proxy_{clip['id']}.mp4"
        count += 1

        db.setdefault("transcodingJobs", []).append({
            "id": f"job_all_{clip['id']}_{int(time.time() * 1000)}",
            "clipId": clip["id"],
            "mediaTitle": clip["title"],
            "originalResolution": "1080p FHD",
            "targetResolution": target_res,
            "progressPercent": 100,
            "statusMessage": f"Proxy Transcoded ({target_res})",
            "isCompleted": True
        })

    save_db(db)

    if json_mode:
        return json.dumps({"success": True, "transcodedCount": count, "targetResolution": target_res}, indent=2)
    return f"\n🚀 BACKGROUND TRANSCODER ENGINE (Python)\n=========================================\n✅ Berhasil mentranscode {count} klip video ke resolusi proxy ({target_res})!\nBeban GPU Timeline berkurang ~75% untuk pemutaran 60 FPS tanpa lag.\n"

def view_proxy_status(json_mode=False):
    db = get_db()
    s = db.get("settings", {})
    jobs = db.get("transcodingJobs", [])

    if json_mode:
        return json.dumps({"settings": s, "jobs": jobs}, indent=2)

    is_proxy = s.get("isProxyModeEnabled", True)
    res = s.get("proxyResolution", "360p Proxy")
    auto_trans = s.get("autoTranscodeOnImport", True)

    out = f"\n⚙️ STATUS LOW-RESOLUTION PROXY ENGINE (Python)\n"
    out += f"=========================================\n"
    out += f"Mode Proxy Preview : {'⚡ AKTIF' if is_proxy else '🎬 NONAKTIF (1080p Direct)'}\n"
    out += f"Target Resolution  : {res}\n"
    out += f"Auto-Transcode     : {'AKTIF' if auto_trans else 'NONAKTIF'}\n"
    out += f"Beban GPU Estimasi : {'15% - 25% (Sangat Ringan)' if is_proxy else '85% - 100% (High-Res Direct)'}\n\n"

    out += f"📋 RIWAYAT TUGAS TRANSCODER LATAR BELAKANG ({len(jobs)} Tugas):\n"
    if not jobs:
        out += "   (Belum ada tugas transcoding)\n"
    else:
        for j in jobs[-5:]:
            status_tag = "DONE" if j.get("isCompleted") else "IN_PROGRESS"
            out += f"   - [{status_tag}] {j.get('mediaTitle')} ({j.get('originalResolution')} -> {j.get('targetResolution')}) - Progress: {j.get('progressPercent')}%\n"
    return out
